Introduce new MovementInput component

gfontenot opened the pull request:
This component replaces the old single-text-field approach for user input in
favor of dynamically adding/removing fields as the user types. This behavior
is _much_ nicer, and means we'll be able to do more things like
auto-completion easier than we otherwise would have.

![movementinput](https://user-images.githubusercontent.com/124499/36643950-403d4468-1a18-11e8-8186-57bea323ed40.gif)

gfontenot committed:
Move elm-specific git-ignore into elm subdirectory

gfontenot committed:
Add simple movement input component

This component is super basic right now, but will be the main interface for
adding movements to a movement group (still not sure what to call this). It
basically replaces the existing behavior of using a single text field for
multiple movements. It's likely that this will be broken down into
sub-components eventually.

gfontenot committed:
Clarify intent by using Movements instead of Model

Model is overly-generic in this case, and we _really_ expect to be working
with List Movement. For the sake of brevity, we can go ahead and typealias
that to Movements, but lets name things accordingly.

gfontenot committed:
Intersperse movement fields with +

This gives the visual indication that the user is entering a complex, instead
of a single movement. It also mimics the existing `foo + bar + baz` behavior
nicely.

gfontenot committed:
Add movements manually via button click

I want to let users type + to add a new movement, but I _also_ want to let
them click a button to manually add a new movement to the complex. The former
behavior exists for speed, the latter is for discoverability.

gfontenot committed:
Track movement index

Previously, when a movement changed, we were just returning that movement
again. That's not actually what we want, though. Instead, in order to handle
changes properly, we need to be able to track which movement changed and edit
it directly.

By tracking the index, we're also able to do things like insert a new set in
the middle of the list by typing `+`.

gfontenot committed:
Update focus when we add a new set

When we're adding a new set, it's always via user interaction, and so we
always want to update the focus into the new field. This is true if we're
hitting the `+` button, or if we've typed "+" to add a new field.

Updating the focus is done through the `Dom` package. The task that it uses to
update the focus can fail, so we need to wrap it in our own function. We don't
really care if the focus fails right now, so we'll just swallow the error.

To make this work, we also need to add a new message, which we'll treat as a
no-op.

gfontenot committed:
Delete movements by deleting text

When the user deletes through an empty field, we want to delete the movement
itself. This will give the impression of deleting the + symbol, which further
enhances the illusion of plain-text entry.

gfontenot committed:
Extract Movement type to new file

This type is literally a type-alias right now, but it's probably worth
extracting it before we move too much further.

gfontenot committed:
Replace old Movement field with MovementInput

This removes the old Movement field from our Main module, and replaces it with
our new MovementInput component, which handles the all of the internal state
on its own.

gabebw commented on elm/Helpers/Events.elm:
```diff
@@ -1,9 +1,21 @@
+            (True, True) -> Json.succeed m
+            (True, False) -> Json.fail "Text not empty"
+            (False, True) -> Json.fail "Wrong key pressed"
+            (False, False) -> Json.fail "Wrong key pressed, text not empay"
```
Typo: "empay"

gfontenot replied:
to be fair the text probably also isn't "empay"

gfontenot replied:
although I suppose I could make that assertion more explicit

gfontenot committed:
Simplify onDeleteEmpty -> onDelete

It appears that this doesn't actually trigger unless the field is empty
anyway. I'm not 100% sure why that is, but given that fact, we can simplify
this to be `onDelete`. Additionally, we can clean up the code a bit so that
the Json operations are easier to follow.

gfontenot committed:
Rename FocusedField to NoOp

This is closer to the standard practice, and is _actually_ what we want. We
don't really care about knowing that the field was focused (yet?), so there's
not a whole lot of point in having an explicitly named message for that.
Instead, we can use a basic NoOp message to indicate that we're ignoring the
message entirely.

gfontenot committed:
Remove MovementInput.main

This was only needed for initial development. It doesn't need to exist now
that I've added this component into the main module.

gfontenot committed:
I was wrong, re-add precondition to onDelete

Nope that was incorrect. This definitely does not get sent only when the field
is empty. So we definitely need to do some checking to ensure that our string
is empty before deleting.

That being said, the way were were handling this before was not great. This
feels a _bit_ better, but still not 100% right. Instead of taking the string,
we'll just take a boolean to use as a precondition. That way, we can offload
the responsibility of checking the string value to the component.

There's still almost certainly a better way to accomplish this, but I can't
see it yet.

gfontenot committed:
Don't strip whitespace from input

This isn't actually what we wanted. We should definitely strip whitespace
eventually, but stripping it here means you can't actually enter any
whitespace. That's not good.

joshuaclayton commented on elm/Swole/Types/Movement.elm:
```diff
@@ -0,0 +1,5 @@
+    ( Movement
+    )
+
+type alias Movement = String
```
I'll "newtype" this in Elm, instead of the type alias, to be a bit more typesafe:

```elm
type Movement
    = Movement String
```

If you do this, be sure to update the export to include the data constructor.

gfontenot replied:
I'm fully expecting this type to become significantly more complex in the future. This current typealias is just a stop-gap for now.

joshuaclayton replied:
That sounds good, especially if the expectation is that it'll change in the (near) future

joshuaclayton commented on elm/Swole/Components/MovementInput.elm:
```diff
@@ -0,0 +1,112 @@
+        in
+           fields ++ [plusButton]
+
+plusLabel : Html Msg
```
For markup that doesn't actually trigger a `Msg`, I'll use the more polymorphic `Msg a` to convey that. Thoughts?

gfontenot replied:
you mean `Html a`? I like that.

joshuaclayton replied:
Yep, exactly. It's nice! Clarifies intent, since it signifies "nothing within this, or any of its children, can trigger a side-effect"

joshuaclayton commented on elm/Swole/Components/MovementInput.elm:
```diff
@@ -0,0 +1,112 @@
+
+movementFields : Movements -> List (Html Msg)
+movementFields movements = case movements of
+    [] -> [movementField (0, "")]
```
Thoughts on extracting this `movementField (0, "")` into a separate function? Maybe `defaultMovementField`, `initialMovementField`, or similar?

joshuaclayton commented on elm/Swole/Components/MovementInput.elm:
```diff
@@ -0,0 +1,112 @@
+updateFocus idx =
+    let
+        checkFocus : Result Dom.Error () -> Msg
+        checkFocus result = case result of
```
You could skip this case statement entirely if you don't plan on using it in the future:

```elm
checkFocus = always NoOp
```

joshuaclayton commented on elm/Swole/Components/MovementInput.elm:
```diff
@@ -0,0 +1,112 @@
+movementsAround : Int -> Movements -> (List Movement, List Movement)
+movementsAround idx movements
+    = movements
+    |> List.map Tuple.second
```
Having looked at this file, there are a significant number of functions that interact with the list of pairs. I'm wondering if there's a more concrete, custom type that could be built, with a handful of functions specific to that type, that better convey the intent here. Specifically, updating movements and deleting movements have a decent amount of complexity.

gfontenot replied:
Yes, for sure. This is also going to be a common pattern, so I'm very interested in finding a more reasonable solution to this.

joshuaclayton replied:
I think for now, I'd build up a type to capture this concept, and think about generalization after it crops up again?

gfontenot replied:
It technically already did crop up again, this is the exact same pattern I'm using inside Main for handling multiple sets. I'm going to move forward with this and open up a new PR trying to move to `SelectList`.

joshuaclayton commented on elm/Main.elm:
```diff
@@ -85,11 +59,32 @@ update msg model =
+        AddSet ->
+            let
+                count = 0
+                newSets = model.setModels ++ [SingleSet.initialModel count]
```
`SingleSet.initialModel 0` is used in a couple of spots in the codebase; is it ever called with a value other than `0`? If not, thoughts on removing that argument?

gfontenot replied:
Yes, definitely. That's an artifact from some old design decisions that I'm planning on refactoring/rewriting relatively soon. I intentionally avoided diving into that rabbit hole during these changes.

joshuaclayton replied:
👍

joshuaclayton commented on elm/Main.elm:
```diff
@@ -85,11 +59,32 @@ update msg model =
+        AddSet ->
+            let
+                count = 0
+                newSets = model.setModels ++ [SingleSet.initialModel count]
```
Is there a way to roll up this into a function on `SingleSet` that captures adding a new set (including no side effects) so the structure would be similar to that of `MovementInputMsg msg`s?

gfontenot replied:
Yeah, I'm almost certainly going to move in that direction. Like with the code above, this is an artifact and can be ignored right now.

joshuaclayton replied:
Cool beans

joshuaclayton commented on elm/Swole/Components/MovementInput.elm:
```diff
@@ -0,0 +1,112 @@
+        MovementChanged (idx, m) ->
+            let
+                (newIdx, newMovements) = splitMovements (idx, m)
+                (before, after) = movementsAround idx movements
```
I think we'd discussed this structure when we talked; would it work here? http://package.elm-lang.org/packages/rtfeldman/selectlist/latest

gfontenot replied:
Yeah I need to look into this. The docs... aren't great, so I'm not 100% sure how to _use_ this, but I'm very interested in removing (or extracting) all of my manual indexing code.

joshuaclayton replied:
Yeah, the concept is that it's a non-empty list with a focus, and the data structure is such that there's a list before the selected item (if it's empty, the first item in the "list" is selected), and there's a list of items after the selected item (if it's also empty, there's only a single member in the list, the selected item).

I'm not sure if the underlying data structure here *requires* one item - if it does, this should work, but it might be outside the scope of this? There are a few functions I ended up having to write myself to make the package more usable:

```elm
selectSlide : (a -> comparable) -> SelectList a -> a -> SelectList a -- this selects a new member, with the understanding there's an accessor function to identify `a`
replaceSelected : SelectList a -> a -> SelectList a -- this swaps out the current selected value
buildSelectList : a -> List a -> SelectList a
deleteSlide : (a -> comparable) -> SelectList a -> a -> SelectList a -- same thing as `selectSlide`, given an accessor function that finds a particular item from `a` and the value you want to delete (`a`), it deletes the item and selects the closest one available. If there's only one value, it doesn't modify
```

joshuaclayton reviewed (commented):
Some small refactoring/pattern opportunities here.

Independent of this PR, thoughts on hooking up `elm-format` with the codebase? It trims off unnecessary parens, among other things, but it'll help ensure formatting consistency.

gfontenot committed:
Refactor updateFocus to remove cast statement

The case statement I was using is functionally equivalent to just using
`always NoOp`, so we should just do that. While we're at it, we can improve
the readability of this with some small helper functions.

gfontenot committed:
Extract default movement field to a function

This makes it clearer that it's a default value.

gfontenot committed:
Make plusLabel polymorphic

We don't care about the message type so lets loosen up the type here.
